crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4444,9 +4444,18 @@ impl<'db> TypeInferenceBuilder<'db> {
                 }
 
                 if return_todo {
+                    self.store_expression_type(tuple_slice, Type::Todo);
+
                     Type::Todo
                 } else {
-                    Type::tuple(self.db, &element_types)
+                    let tuple_type = Type::tuple(self.db, &element_types);
+
+                    // Here, we store the type for the inner `int, str` tuple-expression,
+                    // while the type for the outer `tuple[int, str]` slice-expression is
+                    // stored in the outer `infer_type_expression` call:
+                    self.store_expression_type(tuple_slice, tuple_type);
+
+                    tuple_type
                 }
```

Comment:
nit: I'd prefer to do something like this, so that we only have to call `self.store_expression_type()` once:

```rs
                let ty = if return_todo {
                    Type::Todo
                } else {
                    Type::tuple(self.db, &element_types)
                };

                // Here, we store the type for the inner `int, str` tuple-expression,
                // while the type for the outer `tuple[int, str]` slice-expression is
                // stored in the outer `infer_type_expression` call:
                self.store_expression_type(tuple_slice, ty);

                ty
```